### PR TITLE
Exactness Indicator of Parameters: Precision

### PR DESCRIPTION
**Pull request by berkaysynnada**
## Which issue does this PR close?

Closes #.

## Rationale for this change

Some parameters need to be carried along with their exactness information. That exactness information can be provided by the source, or it can be set during the execution. This implementation which will initially be useful in statistics can also be used in different areas in the future.

If this approach makes sense and is approved by the community, I will subsequently open a new [PR](https://github.com/apache/arrow-datafusion/pull/7793) where its importance on statistics will be seen and some optimizations will be possible.

## What changes are included in this PR?

An enumeration, named Precision, is implemented to define 3 types of exactness variants: Exact, Inexact, and Absent. The enum can be used in a generic way, and some utilities on usize and ScalarValue have also been added.

## Are these changes tested?

Yes, unit tests are added.

### CODE REVIEW

**Comment by alamb** [comment]
This is part of https://github.com/apache/arrow-datafusion/pull/7793, correct? If so, thank you for extracting it to its own PR

**Comment by berkaysynnada** [comment]
> This is part of #7793, correct? If so, thank you for extracting it to its own PR

Yes, it is. #7804 is also another extracted part.

File: datafusion/common/src/stats.rs [hunk 1]
```diff
@@ -17,11 +17,184 @@
...
-use std::fmt::Display;
-
+use crate::ScalarValue;
 use arrow::datatypes::DataType;
 
-use crate::ScalarValue;
+use std::fmt::{self, Debug, Display};
+
+/// To deal with information without exactness guarantees, we wrap it inside a
+/// [`Sharpness`] object to express its reliability.
```
Comment (alamb) [inline datafusion/common/src/stats.rs:26]:
```suggestion
/// Represents a value with a degree of certainty. `Sharpness` is used to 
/// propagate information the precision of statistical values.
```

While writing this I wonder if naming `Sharpness` to  `Precision` might match more closely with what this type represents 🤔

File: datafusion/common/src/stats.rs [hunk 2]
```diff
@@ -17,11 +17,184 @@
...
+#[derive(Clone, PartialEq, Eq, Default)]
+pub enum Sharpness<T: Debug + Clone + PartialEq + Eq + PartialOrd> {
+    Exact(T),
+    Inexact(T),
+    #[default]
+    Absent,
```
Comment (alamb) [inline datafusion/common/src/stats.rs:32]:
```suggestion
    /// The exact value is known 
    Exact(T),
    /// The value is not known exactly, but is likely close to this value
    Inexact(T),
    /// Nothing is known about the value
    #[default]    
    Absent,
```

File: datafusion/common/src/stats.rs [hunk 3]
```diff
@@ -17,11 +17,184 @@
...
+            (Sharpness::Inexact(a), Sharpness::Exact(b))
+            | (Sharpness::Exact(a), Sharpness::Inexact(b))
+            | (Sharpness::Inexact(a), Sharpness::Inexact(b)) => {
+                Sharpness::Inexact(if a >= b { b.clone() } else { a.clone() })
+            }
+            (_, _) => Sharpness::Absent,
+        }
+    }
+
+    /// Demotes the sharpness state to inexact (if present).
```
Comment (alamb) [inline datafusion/common/src/stats.rs:102]:
```suggestion
    /// Demotes the sharpness state from exact to inexact (if present)
```

**Review by alamb — APPROVED** [review]
LGTM -- thank you for extracting this out @berkaysynnada

**Comment by berkaysynnada** [comment]
> LGTM -- thank you for extracting this out @berkaysynnada

Thank you for the reviews. After these 2 PRs are merged, I will resolve the conflicts of #7793 and request a review again.

**Pull request merged**